CSHARP-2171: SCRAM-SHA-256 Support

vincentkam opened the pull request:
Evergreen patch:
https://evergreen.mongodb.com/version/5b23d5422fbabe4163d8e0f4

New PR as discussed.

vincentkam commented on src/MongoDB.Driver.Core/Core/Authentication/Vendored/CryptographyHelpers.cs:
```diff
@@ -0,0 +1,175 @@
+    internal static class CryptographyHelpers
+    {
+        
+        private static readonly string Cryptography_MissingIV
```
Originally in a separate string resources class.

vincentkam commented on src/MongoDB.Driver.Core/Core/Authentication/Vendored/Rfc2898DeriveBytes.cs:
```diff
@@ -0,0 +1,337 @@
+            Initialize();
+        }
+        
+        private static string Format(string resourceFormat, params object[] args)
```
Initially in a separate string resources class.

vincentkam commented on src/MongoDB.Driver.Core/Core/Authentication/Vendored/Rfc2898DeriveBytes.cs:
```diff
@@ -0,0 +1,337 @@
+    
+    internal class Rfc2898DeriveBytes : DeriveBytes
+    {
+        private static readonly string Cryptography_PasswordDerivedBytes_FewBytesSalt 
```
Error messages initially in a separate string resources class, but for simplicity's sake, I've defined them in the class. Additionally, we don't need to provide localizations for error messages.

rstam commented on src/MongoDB.Driver.Core/Core/Authentication/AuthenticationHelper.cs:
```diff
@@ -36,27 +35,34 @@ public static void Authenticate(IConnection connection, ConnectionDescription de
 
             // authentication is currently broken on arbiters
-            if (!description.IsMasterResult.IsArbiter)
+            if (description.IsMasterResult.IsArbiter)
```
I have no objection to these changes, but this whole file could have been left unchanged.

vincentkam replied:
Apologies, initially there were much more significant changes to this file, which then later become unnecessary. In the future, if this happens, I'll back out the insignificant changes as well.

rstam replied:
OK. But none of the diffs in this file change the behavior of the code in any way.

vincentkam replied:
Backing out the changes.

rstam commented on src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs:
```diff
@@ -14,44 +14,53 @@
         }
 
-        internal DefaultAuthenticator(UsernamePasswordCredential credential, IRandomStringGenerator randomStringGenerator)
+        internal DefaultAuthenticator(UsernamePasswordCredential credential,
```
If you're going to put each parameter on a separate line put the first one on a new line also.

vincentkam replied:
Ack, fixed.

rstam commented on src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs:
```diff
@@ -60,30 +69,86 @@ public void Authenticate(IConnection connection, ConnectionDescription descripti
-                return new ScramSha1Authenticator(_credential, _randomStringGenerator);
-            }
-            else
+            return isMasterCommand.Merge(SaslSupportedMechanismsQuery); 
```
I think we can just call CreateSaslSupportedMechsRequest and get rid of the SaslSupportedMechanismsQuery property and backing Lazy field entirely.

vincentkam replied:
Fixed.

rstam commented on src/MongoDB.Driver.Core/Core/Connections/IsMasterResult.cs:
```diff
@@ -206,6 +206,29 @@ public EndPoint Me
+        /// <value>
+        /// Whether SaslSupportedMechs was part of the isMaster response.
+        /// </value>
+        public bool SaslSupportedMechsReceived
```
By our standard naming conventions this would normally be called HasSaslSupportedMechs

vincentkam replied:
Fixed.

rstam commented on src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs:
```diff
@@ -60,30 +69,86 @@ public void Authenticate(IConnection connection, ConnectionDescription descripti
+                mergedIsMasterResult, 
+                description.BuildInfoResult);
+
+            CreateAuthenticator(connection, mergedDescription).Authenticate(connection, mergedDescription, cancellationToken);
```
When I first read this line I thought "weird we just create an authenticator and do nothing with it".

Can you introduce an intermediate variable and write this as two lines. It's too easy to miss the call to Authenticate buried farther to the right.

vincentkam replied:
Sure, no problem. I initially chose this style for brevity's sake (and also because the line fit almost entirely within 120 characters, where one can reasonably expect "important" stuff to happen). But I have no issues introducing the interim variable.

Fixed!

rstam commented on src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs:
```diff
@@ -60,30 +69,86 @@ public void Authenticate(IConnection connection, ConnectionDescription descripti
+            if (description.IsMasterResult.SaslSupportedMechsReceived 
+                || !Feature.ScramSha256Authentication.IsSupported(description.ServerVersion)) 
+            {
+                CreateAuthenticator(connection, description).Authenticate(connection, description, cancellationToken);
```
Can we have a single call to CreateAuthenticator (and Authenticate) at the bottom of this method?

What should be in the if statement should be the optional call to isMaster (if necessary).

    if (!description.IsMasterResult.HasSaslSupportedMechs &&
        Feature.ScramSha256Authentication.IsSupported(description.ServerVersion))
    {
        // run an isMaster command and merge the saslSupportedMechs into the description
    }

    var authenticator = CreateAuthenticator(connection, description);
    authenticator.Authenticate(connection, description, cancellationToken)

vincentkam replied:
Discussed offline, fixed.

rstam commented on src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs:
```diff
@@ -60,30 +69,86 @@ public void Authenticate(IConnection connection, ConnectionDescription descripti
+            if (description.IsMasterResult.SaslSupportedMechsReceived 
+                || !Feature.ScramSha256Authentication.IsSupported(description.ServerVersion))
+            {
+                await CreateAuthenticator(connection, description)
```
See comment in the sync version.

vincentkam replied:
Fixed.

rstam commented on src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs:
```diff
@@ -60,30 +69,86 @@ public void Authenticate(IConnection connection, ConnectionDescription descripti
+             
+            return Feature.ScramSha1Authentication.IsSupported(description.ServerVersion)
+                    ? (IAuthenticator) new ScramSha1Authenticator(_credential, _randomStringGenerator)
+                    : new MongoDBCRAuthenticator(_credential);
```
This method is still confusing. But I'm not sure how to make it less confusing.

One question: on line 146 what if saslSupportedMechs does NOT contain "SCRAM-SHA-1"? Do we still create a ScramSha1Authenticator anyway?

I wonder if the intent of the negotiation should really be this:

    if (saslSupportedMechanisms.Any())
    {
        if (saslSupportedMechanisms.Contains("SCRAM-SHA-256"))
        {
            return new ScramSha256Authenticator(...);
        }

        if (saslSupportedMechanisms.Contains("SCRAM-SHA-1"))
        {
            return new ScramSha1Authenticator(...);
        }

        throw new MongoAuthenticationException("Driver has no supported mechanism in common with the server.");
    }
    else
    {
        return Feature.ScramSha1Authentication.IsSupported(description.ServerVersion)
            ? (IAuthenticator) new ScramSha1Authenticator(_credential, _randomStringGenerator)
             : new MongoDBCRAuthenticator(_credential);
    }

vincentkam replied:
You're right, this part is a bit confusing, and I had to got back to the spec multiple times.
https://github.com/mongodb/specifications/blob/master/source/auth/auth.rst#defaults
"If a saslSupportedMechs field was present in the isMaster results for mechanism negotiation then it MUST be inspected to select a default mechanism."
"If SCRAM-SHA-256 is present in the list of mechanisms, then it MUST be used as the default; otherwise, SCRAM-SHA-1 MUST be used as the default, regardless of whether SCRAM-SHA-1 is in the list."
Adding comments to code so that this is clearer. 

Additionally, as I reviewed the spec one more time, I realized I made a mistake, so I changed the check in the if statement to ensure that we select SCRAM-SHA-1 as a default IFF there was a "saslSupportedMechs" in the response. So thanks for pointing out this bit was confusing! :)

rstam replied:
So still no exception if the driver doesn't support any of the mechanisms that the server returns in saslSupportedMechanisms?

rstam replied:
I read the part of the spec you quoted and you are right that SCRAM-SHA-1 should be selected EVEN IF THE SERVER DOESN'T REPORT THAT IT SUPPORTS IT.

That makes no sense to me, but let's follow the spec.

vincentkam replied:
Yes, I thought this was odd too! Glad to hear you agree :) Also agreed that let's go with the spec for now.

rstam commented on src/MongoDB.Driver.Core/Core/Authentication/IAuthenticator.cs:
```diff
@@ -49,5 +50,12 @@ public interface IAuthenticator
         Task AuthenticateAsync(IConnection connection, ConnectionDescription description, CancellationToken cancellationToken);
+        
+        /// <summary>
+        /// Optionally customizes isMaster command with additional attributes.
```
Remove "with additional attributes".

In theory it can customize the isMaster command in any way necessary.

vincentkam replied:
Fixed.

rstam commented on src/MongoDB.Driver.Core/Core/Authentication/IAuthenticator.cs:
```diff
@@ -49,5 +50,12 @@ public interface IAuthenticator
+        /// Optionally customizes isMaster command with additional attributes.
+        /// </summary>
+        /// <param name="isMasterCommand">Initial isMaster command.</param>
+        /// <returns>Optionally mutated isMaster command</returns>
```
Remember to end all doc comments with a period.

vincentkam replied:
Fixed

rstam commented on src/MongoDB.Driver.Core/Core/Authentication/ScramSha1Authenticator.cs:
```diff
@@ -16,18 +16,21 @@
+using MongoDB.Driver.Core.Authentication.Vendored;
 using MongoDB.Driver.Core.Connections;
 using MongoDB.Driver.Core.Misc;
+using Rfc2898DeriveBytes = MongoDB.Driver.Core.Authentication.Vendored.Rfc2898DeriveBytes;
```
This is acceptable, though I'm curious why you didn't just do:

    using MongoDB.Driver.Core.Authentication.Vendored;

rstam replied:
As per our discussion on Slack this is because we want to use OUR vendored version of Rfc2898DeriveBytes instead of the one in .NET Framework.

Let's put a blank line to separate this using statement from the others (it serves a completely different purpose) and add a comment, something like:

    // use our vendored version of Rfc2898DerivedBytes instead of the standard one
    using Rfc2898DeriveBytes = MongoDB.Driver.Core.Authentication.Vendored.Rfc2898DeriveBytes;

vincentkam replied:
Sure, setting off the using statement from the rest of the using statement with an extra new line sounds like a good idea. As for the comment, I feel like I already have a comment to the effect? I can change the wording if you'd like.

vincentkam replied:
Added line break to ScramSha1Authenticator. Added both a line break and the additional comment to ScramSha256Authenticator

vincentkam replied:
Additionally, changed the wording to be closer to your format.

rstam replied:
My comment above was written *before* you had added any comment to the file. The comment you added to the file is fine.

vincentkam replied:
Sorry, I thought I pushed a version of the file with a comment on it before your comment, but perhaps it was just a timing mismatch--or another bit a wonkiness that seems to be plaguing the PRs for 2171! Apologies for the confusion--I think we're on the same page though :)

rstam commented on src/MongoDB.Driver.Core/Core/Connections/ConnectionInitializer.cs:
```diff
@@ -28,7 +31,7 @@ namespace MongoDB.Driver.Core.Connections
     internal class ConnectionInitializer : IConnectionInitializer
     {
-        private readonly BsonDocument _clientDocument;
+        internal readonly BsonDocument _clientDocument;
```
internal???

If this was done just for testing purposes normally we'd prefer to use reflection to access it.

vincentkam replied:
Hm, I thought part of the reason we had the InternalsVisibleToAttribute was so that we didn't need to use reflection in the test projects? Discussing offline...

vincentkam replied:
Fixed.

rstam commented on src/MongoDB.Driver.Core/Core/Connections/ConnectionInitializer.cs:
```diff
@@ -64,15 +67,24 @@ public ConnectionDescription InitializeConnection(IConnection connection, Cancel
         }
 
-        public async Task<ConnectionDescription> InitializeConnectionAsync(IConnection connection, CancellationToken cancellationToken)
+        private BsonDocument CreateInitialIsMasterCommand(IReadOnlyList<IAuthenticator> authenticators)
```
Move this further down in the file with the rest of the private methods.

vincentkam replied:
Fixed.

rstam commented on src/MongoDB.Driver.Core/Core/Connections/ConnectionInitializer.cs:
```diff
@@ -64,15 +67,24 @@ public ConnectionDescription InitializeConnection(IConnection connection, Cancel
+        }
+
+        public async Task<ConnectionDescription> InitializeConnectionAsync(IConnection connection,
+            CancellationToken cancellationToken)
```
Either all parameters on one line or each parameter on a separate line (including the first one).

vincentkam replied:
Fixed.

rstam commented on src/MongoDB.Driver.Core/Core/Connections/ConnectionInitializer.cs:
```diff
@@ -64,15 +67,24 @@ public ConnectionDescription InitializeConnection(IConnection connection, Cancel
 
             var buildInfoProtocol = CreateBuildInfoProtocol();
-            var buildInfoResult = new BuildInfoResult(await buildInfoProtocol.ExecuteAsync(connection, cancellationToken).ConfigureAwait(false));
+            var buildInfoResult =new BuildInfoResult(await buildInfoProtocol.ExecuteAsync(connection, cancellationToken).ConfigureAwait(false));
```
This is a spurious white space diff.

vincentkam replied:
Fixed.

rstam commented on src/MongoDB.Driver.Core/Core/Connections/ConnectionInitializer.cs:
```diff
@@ -98,53 +110,28 @@ private CommandWireProtocol<BsonDocument> CreateBuildInfoProtocol()
-
-        private ConnectionDescription UpdateConnectionIdWithServerValue(ConnectionDescription description, BsonDocument getLastErrorResult)
+        private ConnectionDescription UpdateConnectionIdWithServerValue(ConnectionDescription description,
+            BsonDocument getLastErrorResult)
```
All on one line or each on a separate line...

vincentkam replied:
One of my IDEs is definitely enforcing an auto-format rule that I need to turn off...apologies, and fixed.

rstam commented on src/MongoDB.Driver.Core/Core/Connections/IsMasterHelper.cs:
```diff
@@ -0,0 +1,93 @@
+        
+        internal static BsonDocument CustomizeCommand(BsonDocument command, IReadOnlyList<IAuthenticator> authenticators)
+        {
+            // TODO: Always customize the initial isMaster command when support for multiple authenticators is removed
```
Well not always. There still might be zero authenticators.

vincentkam replied:
True. Nixed the comment.

rstam commented on src/MongoDB.Driver.Core/Core/Connections/ConnectionInitializer.cs:
```diff
@@ -13,13 +13,16 @@
 using MongoDB.Driver.Core.Authentication;
 using MongoDB.Driver.Core.Misc;
 using MongoDB.Driver.Core.WireProtocol;
+using static MongoDB.Driver.Core.Connections.IsMasterHelper;
```
Didn't we agree *not* to do this?

vincentkam replied:
We did agree not to, my apologies. I think I made the change to a different branch by accident.

rstam commented on src/MongoDB.Driver.Core/Core/Connections/ConnectionInitializer.cs:
```diff
@@ -64,15 +67,24 @@ public ConnectionDescription InitializeConnection(IConnection connection, Cancel
+        {
+            var command = CreateCommand();
+            AddClientDocumentToCommand(command, _clientDocument);
+            return CustomizeCommand(command, authenticators);
```
These method calls *look* like calls to method in this same class, but they're not...

vincentkam replied:
Due to a misplaced "using static." Fixed.

rstam commented on src/MongoDB.Driver.Core/Core/Connections/IsMasterResult.cs:
```diff
@@ -382,5 +405,6 @@ public ReplicaSetConfig GetReplicaSetConfig()
 
             return new ReplicaSetConfig(members, name, primary, version);
         }
+        
```
Spurious white space diff.

rstam commented on src/MongoDB.Driver.Core/Core/Misc/Feature.cs:
```diff
@@ -58,6 +58,7 @@ public class Feature
         private static readonly Feature __partialIndexes = new Feature("PartialIndexes", new SemanticVersion(3, 2, 0));
         private static readonly ReadConcernFeature __readConcern = new ReadConcernFeature("ReadConcern", new SemanticVersion(3, 2, 0));
         private static readonly Feature __scramSha1Authentication = new Feature("ScramSha1Authentication", new SemanticVersion(3, 0, 0));
+        private static readonly Feature __scramSha256Authentication = new Feature("ScramSha256Authentication", new SemanticVersion(4, 0, 0, "rc0"));
```
A better alternative to "rc0" would be "".

"" means ALL pre-release versions of 4.0.0 (including latest)

"rc0" is more specific and means only "rc0" and newer.

Because of the way the server versions continuous integration builds the version number of latest will be LESS THAN "rc0", which would result in us thinking that SHA256 is NOT supported in latest.

vincentkam replied:
Concerns about "" being a non-obvious way to express what we want. Discussed offline, change made. Good catch! Missed that subtlety regarding server versions.

rstam commented on tests/MongoDB.Driver.Core.Tests/Core/Connections/ConnectionInitializerTests.cs:
```diff
@@ -98,9 +98,10 @@ public ConnectionInitializerTests()
 
         [Fact]
-        public void CreateIsMasterCommand_should_return_expected_result()
+        public void AddClientDocumentToIsMasterCommand_with_subject_client_document_should_return_expected_result()
```
Isn't this testing IsMasterHelper?

This file is testing ConnectionInitializer methods.

@jyemin do we just let this be in order to commit as soon as possible?

vincentkam replied:
True, the CreateIsMasterCommand used to live in ConnectionInitializer. Separate test class added.

rstam replied:
But you didn't move this test method to the new class?

vincentkam replied:
I moved `AddClientDocumentToIsMasterCommand_with_custom_document_should_return_expected_result` to the new class, since that's strictly testing the helper method in IsMasterHelper.

However,  'AddClientDocumentToIsMasterCommand_with_subject_client_document_should_return_expected_result' technically needs a ConnectionInitializer in order to properly run since it's ensuring that a ConnectionInitalizer's _clientDocument is properly added to an isMaster command. It initially looked like this: 
```
        [Fact]
        public void CreateIsMasterCommand_should_return_expected_result()
        {
            var result = _subject.CreateIsMasterCommand();

            var names = result.Names.ToList();
            names.Count.Should().Be(2);
            names[0].Should().Be("isMaster");
            names[1].Should().Be("client");
            result[0].Should().Be(1);
            var clientDocument = result[1].AsBsonDocument;
            var clientDocumentNames = clientDocument.Names.ToList();
            clientDocumentNames.Count.Should().Be(4);
            clientDocumentNames[0].Should().Be("application");
            clientDocumentNames[1].Should().Be("driver");
            clientDocumentNames[2].Should().Be("os");
            clientDocumentNames[3].Should().Be("platform");
            clientDocument["application"]["name"].AsString.Should().Be("test");
            clientDocument["driver"]["name"].AsString.Should().Be("mongo-csharp-driver");
            clientDocument["driver"]["version"].BsonType.Should().Be(BsonType.String);
        }
```

We could remove this test since it's not testing anything that isn't already tested by `AddClientDocumentToIsMasterCommand_with_custom_document_should_return_expected_result`, but I suppose right now it does function as a pseudo-integration test: I could move it IsMasterHelperTests if you'd like.

rstam replied:
Well according to the name of this method you are testing the AddClientDocumentToCommand method, and that's a method on IsMasterHelper, not ConnectionInitializer.

vincentkam replied:
Fair enough, moving the test!

rstam commented on src/MongoDB.Driver.Core/Core/Authentication/DefaultAuthenticator.cs:
```diff
@@ -32,7 +38,7 @@ public class DefaultAuthenticator : IAuthenticator
         // constructors
         /// <summary>
-        /// Initializes a new instance of the <see cref="DefaultAuthenticator"/> class.
+        /// Initializes a new instance of the <see cref="DefaultAuthenticator" /> class.
```
Don't worry about it... but this is another example of a spurious diff.

vincentkam replied:
Fixed, sorry. Might be my IDE getting uppity, unsure. I'll look into it.

vincentkam commented:
I think it may be flaking out a little bit as well. No worries on the stream of e-mails :)
I'm going to try to address the last couple of issues you mentioned in your last set of comments.

vincentkam commented:
Squashed, rebased, updated update. Evergreen is also green! https://evergreen.mongodb.com/version/dot_net_driver_dff7de8a1fe9f0bf5e40a4d5e5ebd35b03ab8873
